security/pkg/k8s/tokenreview/client.go

```diff
@@ -0,0 +1,139 @@
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	k8sauth "k8s.io/api/authentication/v1"
+)
+
+type specForJWTValidationRequest struct {
+	Token string `json:"token"`
+}
+
+type jwtValidationRequest struct {
+	APIVersion string                      `json:"apiVersion"`
+	Kind       string                      `json:"kind"`
+	Spec       specForJWTValidationRequest `json:"spec"`
+}
+
+// Client is the client to call the TokenReview API.
+type Client struct {
+	url         string
+	caCert      []byte
+	reviewerJWT string
+}
+
+// NewClient creates a new tokenReviewClient.
+func NewClient(url string, caCert []byte, reviewerJWT string) *Client {
+	return &Client{
+		url:         url,
+		caCert:      caCert,
+		reviewerJWT: reviewerJWT,
+	}
+}
+
+// Review calls the K8s TokenReview API to verify a JWT, and returns the
+// <namespace>:<serviceaccountname> in the JWT if successful.
+// targetJWT: the target JWT to be verified.
+func (c *Client) Review(targetJWT string) (string, error) {
+	resp, err := c.reviewJWTAtK8sAPIServer(targetJWT)
+	if err != nil {
+		return "", fmt.Errorf("failed to get a token review response: %v", err)
+	}
+	// Check that the JWT is valid
+	if !(resp.StatusCode == http.StatusOK ||
+		resp.StatusCode == http.StatusCreated ||
+		resp.StatusCode == http.StatusAccepted) {
+		return "", fmt.Errorf("invalid review response status code %v", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read from the response body: %v", err)
+	}
+	tokenReview := &k8sauth.TokenReview{}
+	err = json.Unmarshal(bodyBytes, tokenReview)
+	if err != nil {
+		return "", fmt.Errorf("unmarshal response body returns an error: %v", err)
+	}
+	if tokenReview.Status.Error != "" {
+		return "", fmt.Errorf("the TokenReview server returns error status: %v", tokenReview.Status.Error)
+	}
+	if !tokenReview.Status.Authenticated {
+		return "", fmt.Errorf("the TokenReview server authentication failed")
+	}
+	inServiceAccountGroup := false
+	for _, group := range tokenReview.Status.User.Groups {
+		if group == "system:serviceaccounts" {
+			inServiceAccountGroup = true
+			break
+		}
+	}
+	if !inServiceAccountGroup {
+		return "", fmt.Errorf("the JWT is not a service account")
+	}
+	// "username" is in the form of system:serviceaccount:{namespace}:{service account name}",
+	// e.g., "username":"system:serviceaccount:default:example-pod-sa"
+	subStrings := strings.Split(tokenReview.Status.User.Username, ":")
+	if len(subStrings) != 4 {
+		return "", fmt.Errorf("invalid username field in the token review result: %s", tokenReview.Status.User.Username)
+	}
+
+	return subStrings[2] + ":" + subStrings[3], nil
+}
+
+// reviewJWTAtK8sAPIServer reviews the target JWT at k8s API server.
+// targetJWT: the target JWT to be verified.
+func (c *Client) reviewJWTAtK8sAPIServer(targetJWT string) (*http.Response, error) {
```

Comment:
Please replace this function with the library function in #10672.

Reply:
The function in #10672 that reviews a JWT at API server is reviewServiceAccountAtK8sAPIServer().  When calling ValidateK8sJwt(), ValidateK8sJwt() will call reviewServiceAccountAtK8sAPIServer().